graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,327 @@
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecution(InstrumentationExecutionParameters parameters) {
+    return new InstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onDispatched(CompletableFuture<ExecutionResult> completableFuture) {}
+
+      @Override
+      public void onCompleted(ExecutionResult executionResult, Throwable throwable) {
+        CacheControlState state = parameters.getInstrumentationState();
+
+        // Attach the policy to the context object
+        state.overallPolicy.maybeAsString().ifPresent(s -> parameters.getGraphQLContext().put(CONTEXT_KEY, s));
+      }
+    };
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginField(
+      InstrumentationFieldParameters parameters) {
+    CacheControlState state = parameters.getInstrumentationState();
+    CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+    boolean inheritMaxAge = false;
+
+    GraphQLUnmodifiedType unwrappedFieldType =
+        GraphQLTypeUtil.unwrapAll(parameters.getExecutionStepInfo().getType());
+
+    // There's no way to set a cacheControl directive on the _entities field or
+    // the _Entity union in SDL. Instead, we can determine the possible concrete
+    // types from the representations arguments and select the most restrictive
+    // cache policy from those types.
+
+    if (unwrappedFieldType.getName().equals(_Entity.typeName)) {
+      Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+      if (representations instanceof List) {
+        typesFromEntitiesArgument(
+            representations, parameters.getExecutionContext().getGraphQLSchema())
+            .stream()
+            .map(type -> CacheControlDirective.fromDirectiveContainer((GraphQLDirectiveContainer) type))
+            .filter(Optional::isPresent)
+            .forEach(directive -> fieldPolicy.restrict(directive.get()));
+      }
+    } else if (unwrappedFieldType instanceof GraphQLCompositeType) {
+
+      // Cache directive on the return type of this field if it's a composite type
+
+      Optional<CacheControlDirective> directive = CacheControlDirective.fromDirectiveContainer((GraphQLDirectiveContainer) unwrappedFieldType);
+
+      if (directive.isPresent()) {
+        fieldPolicy.replace(directive.get());
+        inheritMaxAge = directive.get().getInheritMaxAge();
+      }
+    }
+
+    // Cache directive on the field itself
+
+    Optional<CacheControlDirective> fieldDirective = CacheControlDirective.fromDirectiveContainer(parameters.getField());
+    if (fieldDirective.isPresent()) {
+      CacheControlDirective directive = fieldDirective.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the
+      // default max age in the next step. This does allow setting the cache
+      // scope though.
+      //
+      // Note: setting both maxAge and inheritMaxAge doesn't make sense.
```

Comment:
Turns out I was wrong about this. Having `maxAge` and `inheritMaxAge` at the same time is valid:

From the [TS implementation](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/plugin/cacheControl/index.ts#L155):
```
              // Note that specifying `@cacheControl(inheritMaxAge: true)` on a
              // field whose return type defines a `maxAge` gives precedence to
              // the type's `maxAge`. (Perhaps this should be some sort of
              // error.)
```

Reply:
copied the comment from the TS implementation